working/augmentation-libraries/feature-specification.md

```diff
@@ -622,311 +626,97 @@ It is a compile-time error if:
+augment class C {
+  int x = 3;
 
+  static int y = 4;
+}
 ```
-// main.dart
-part 'a.dart';
-part 'c.dart';
 
-enum A {
-  first,
-  second.custom(1);
+Variable declarations themselves can't be directly augmented, with the exception
+of abstract fields.
 
-  final int b;
+*A variable declaration implicitly has code for the synthesized getter and
+setter that access and modify the underlying backing storage. Since we don't
+allow augmentations to replace code, that implies that augmentations can't
+change variables. So we don't allow them to be augmented.*
 
-  const A() : b = 0;
+#### Abstract fields
 
-  const A.custom(this.b);
-}
-}
+Dart supports `abstract` field declarations. They are syntactic sugar for
+declaring an abstract getter with an optional abstract setter if the variable is
+non-final. They don't actually declare a variable with any backing storage.
 
-// a.dart
-part of 'main.dart';
-part 'b.dart';
+Because abstract variables are effectively abstract getter and setter
+declarations, they can be augmented and used in augmentations just like function
+declarations:
 
-augment enum A {
-  third;
+*Examples:*
 
-  /// Some doc comment
-  augment first; // This is still `first` in values.
+```dart
+class C {
+  // Augment an abstract variable with a getter:
+  abstract final int a;
+  augment int get a => 1; // OK.
 
-  @someAnnotation
-  augment second; // Don't repeat the argument list, original is used.
-}
+  // Augment an abstract variable with a getter and setter:
+  abstract int b;
+  augment int get b => 1; // OK.
+  augment set b(int value) {} // OK.
 
-// b.dart
-part of 'a.dart';
+  // Augment a getter with an abstract variable:
+  int get c;
 
-augment enum A {
-  fourth;
+  @someMetadata
+  augment abstract final int c; // (Not very useful, but valid.)
 }
+```
 
-// c.dart
-part of 'main.dart';
-
-augment enum A {
-  fifth;
+### Augmenting enum members
 
-  // Error, enum value augmentations cannot have an explicit constructor
-  // invocation.
-  augment third.custom(3);
-}
-```
+An augmentation of an enum type can add new members to the enum, including new
+enum values. Enum values are appended in augmentation application order.
 
-Then `A.values` is `[A.first, A.second, A.third, A.fourth, A.fifth]`.
+Enum values themselves can't be augmented since they are essentially constant
+variables and variables can't be augmented.
 
-It is a compile-time error if:
+It's a **compile-time error** if:
 
-*   An augmenting getter is defined for an enum value. _An enum value
-    counts as a constant variable._
-*   An enum value augmentation provides an explicit constructor invocation.
+*   A declaration inside an augmenting enum declaration has the name `values`,
+    `index`, `hashCode`, or `==`. *It has always been an error for an enum
+    declaration to declare a member named `index`, `hashCode`, `==`, or
+    `values`, and this rule just clarifies that this error is applicable for
+    augmenting declarations as well.*
```

Comment:
(We should be lifting that rule up to the level of "augmentation chains" or "complete declarations" and say that the *collected* declarations must not contain these members.)

Reply:
Does the distinction matter? For the collected declarations to contain one of these members, either the introductory declaration declares one (already an error) or one of the augmentations does (an error here).

Also, in general we state that if the result of applying augmentations leads to a combined declaration with errors then those errors are reported. (For example, two augmentations adding members with the same name and conflicting signatures.)